File: src/utils/patch.rs
use core::ffi::c_void;
use core::mem::{self, transmute};

use crate::mkb;
use crate::utils::ppc;

pub unsafe fn clear_dc_ic_cache(ptr: *mut c_void, size: usize) {
    mkb::DCFlushRange(ptr as *mut _, size as u32);
    mkb::ICInvalidateRange(ptr as *mut _, size as u32);
}

pub unsafe fn write_branch(ptr: *mut u32, destination: *mut c_void) -> u32 {
    let branch: u32 = 0x48000000; // b
    write_branch_main(ptr, destination, branch)
}

pub unsafe fn write_branch_bl(ptr: *mut u32, destination: *mut c_void) -> u32 {
    let branch: u32 = 0x48000001; // bl
    write_branch_main(ptr, destination, branch)
}

pub unsafe fn write_blr(ptr: *mut u32) -> u32 {
    write_word(ptr, 0x4e800020)
}

unsafe fn write_branch_main(ptr: *mut u32, destination: *mut c_void, mut branch: u32) -> u32 {
    let delta = destination as u32 - ptr as u32;

    branch |= delta & 0x03FFFFFC;

    let p = ptr as *mut u32;
    let orig_word = *p;
    *p = branch;

    clear_dc_ic_cache(ptr as *mut _, size_of::<u32>());

    orig_word
}

pub unsafe fn write_word(ptr: *mut u32, data: u32) -> u32 {
    let p = ptr as *mut u32;
    let orig_word = *p;
    *p = data;
    clear_dc_ic_cache(ptr as *mut _, mem::size_of::<u32>());

    orig_word
}

pub unsafe fn write_nop(ptr: *mut u32) -> u32 {
    write_word(ptr, 0x60000000)
}

pub unsafe fn hook_function_internal(func: *mut c_void, dest: *mut c_void) {
    // Branch directly to the destination function from the original function,
    // leaving no option to call the original function
    let instructions = func as *mut u32;
    write_branch(instructions, dest);
}

pub unsafe fn hook_function(
    func: *mut u32,
    dest: *mut u32,
    tramp_instrs: &mut [u32; 2],
    tramp_dest: &mut *const c_void,
) {
    if (*func & ppc::B_OPCODE_MASK) == ppc::B_OPCODE {
        // Func has been hooked already, chain the hooks

        // Compute dest currently branched to
        let mut old_dest_offset = *func & ppc::B_DEST_MASK;
        // Sign extend to make it actually a s32
        if (old_dest_offset & 0x02000000) != 0 {
            old_dest_offset |= 0xFC000000;
        }
        let old_dest = func as u32 + old_dest_offset;

        // Hook to our new func instead
        write_branch(func, dest as *mut _);

        // Use the old hooked func as the trampoline dest
        *tramp_dest = transmute(old_dest);
    } else {
        // Func has not been hooked yet
        // Original instruction
        tramp_instrs[0] = *func;
        clear_dc_ic_cache(tramp_instrs.as_ptr() as *mut _, size_of::<u32>());

        // Branch to original func past hook
        write_branch(
            &mut tramp_instrs[1] as *mut u32,
            transmute(func.offset(1) as *const _),
        );

        // The function pointer to run as the original function is the addr of the trampoline
        // instructions array
        *tramp_dest = transmute(tramp_instrs.as_ptr());

        // Write actual hook
        write_branch(func, dest as *mut _);
    }
}

File: src/utils/patch.c
#include "utils/patch.h"

static void clear_dc_ic_cache(void *ptr, u32 size) {
    mkb_DCFlushRange(ptr, size);
    mkb_ICInvalidateRange(ptr, size);
}

u32 Patch_WriteB(void *ptr, void *destination) {
    u32 branch = 0x48000000;  // b
    return Patch_WriteBranchMain(ptr, destination, branch);
}

u32 Patch_WriteBL(void *ptr, void *destination) {
    u32 branch = 0x48000001;  // bl
    return Patch_WriteBranchMain(ptr, destination, branch);
}

u32 Patch_WriteBLR(void *ptr) {
    return Patch_WriteWord(ptr, 0x4e800020);
}

u32 Patch_WriteBranchMain(void *ptr, void *destination, u32 branch) {
    u32 delta = (u32)destination - (u32)ptr;

    branch |= (delta & 0x03FFFFFC);

    u32 *p = (u32 *)ptr;
    u32 orig_word = *p;
    *p = branch;

    clear_dc_ic_cache(ptr, sizeof(u32));

    return orig_word;
}

u32 Patch_WriteWord(void *ptr, u32 data) {
    u32 *p = (u32 *)ptr;
    u32 orig_word = *p;
    *p = data;
    clear_dc_ic_cache(ptr, sizeof(u32));

    return orig_word;
}

u32 Patch_WriteNop(void *ptr) {
    return Patch_WriteWord(ptr, 0x60000000);
}

void Patch__HookFunctionReplace(void *func, void *dest) {
    // Branch directly to the destination function from the original function,
    // leaving no option to call the original function
    u32 *instructions = (u32 *)func;
    Patch_WriteB(&instructions[0], dest);
}

void Patch__HookFunctionTramp(void *func, void *dest, u32 *tramp_instrs, void **tramp_dest) {
    u32 *func_instrs = (u32 *)func;

    constexpr u32 B_OPCODE_MASK = 0xFC000000;
    constexpr u32 B_OPCODE = 0x48000000;
    constexpr u32 B_DEST_MASK = 0x03FFFFFC;

    if ((func_instrs[0] & B_OPCODE_MASK) == B_OPCODE) {
        // Func has been hooked already, chain the hooks

        // Compute dest currently branched to
        u32 old_dest_offset = func_instrs[0] & B_DEST_MASK;
        // Sign extend to make it actually a s32
        if (old_dest_offset & (0x02000000)) {
            old_dest_offset |= 0xFC000000;
        }
        u32 old_dest = (u32)func + old_dest_offset;

        // Hook to our new func instead
        Patch_WriteB(&func_instrs[0], dest);

        // Use the old hooked func as the trampoline dest
        *tramp_dest = (void *)old_dest;

    } else {
        // Func has not been hooked yet

        // Original instruction
        tramp_instrs[0] = func_instrs[0];
        clear_dc_ic_cache(tramp_instrs, sizeof(u32));

        // Branch to original func past hook
        Patch_WriteB(&tramp_instrs[1], &func_instrs[1]);

        // The function pointer to run as the original function is the addr of the trampoline
        // instructions array
        *tramp_dest = tramp_instrs;

        // Write actual hook
        Patch_WriteB(&func_instrs[0], dest);
    }
}
